torch/csrc/dynamo/eval_frame.c

```diff
@@ -545,14 +562,13 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
     }
     return (PyObject*)e->code;
   }
-  if (unlikely(guard_fail_hook != NULL)) {
-    PyObject* r = call_guard_fail_hook(guard_fail_hook, e, index, f_locals);
-    if (r == NULL) {
-      return NULL;
-    }
-    Py_DECREF(r);
-  }
-  return lookup(e->next, frame, e, index + 1);
+  // valid == False
+  PyObject* fail_code_part = PyTuple_GetItem(result, 1);
+  Py_INCREF(fail_code_part);
+  Py_DECREF(result);
+  PyObject* lookup_result = lookup(e->next, frame, e, index + 1, code_part);
+  *code_part = fail_code_part;
```

Comment:
Defensive programming demands you assert code_part not NULL here.

Reply:
Wait, why?

Reply:
Well, (1) you didn't give any precondition at the function doc about `code_part`, and it's common that these optional arguments accept null when you don't want an argument, (2) the null test is cheap, just do it.

Reply:
Actually, NULL is valid here? let me make a comment to that.

Reply:
I'm talking about `code_part`, not `fail_code_part`. You should probably rename your variables. If `code_part` is null you'll segfault.

Reply:
No I won't. In fact, it's always NULL going into here?

Or maybe I am misunderstanding you?

https://godbolt.org/z/9Ys4dYhrc

Reply:
You are dereferencing `code_part` to do the assignment here, if `code_part == NULL`, then `*code_part` will segfault.